include/device.h

```diff
@@ -389,6 +389,109 @@ void device_busy_set(struct device *busy_dev);
  */
 void device_busy_clear(struct device *busy_dev);
 
+/**
+ * @brief Device link type codes.
+ */
+enum device_link_type {
+	DEVICE_LINK_TYPE_CLOCK,
+	DEVICE_LINK_TYPE_DMA,
+};
+
+/**
+ * @brief Static device link definition (In ROM)
+ *
+ * @param dev the device linking to another one
+ * @param link_to the link target
+ * @param link_type the type of the link
+ * @param link_index the index within the same link type
+ */
+struct device_link {
+	const struct device *dev;
+	const char *link_to;
+	enum device_link_type link_type;
+	u32_t link_index;
```

Comment:
I have no problem with this, but I'd expect the entire reason for this (instead of simply storing the linked name in the device config) would be so something that doesn't necessarily know the vendor specific ordering can access it.  

For example, without a type field a clock tree walk would have to know what each device has in each of it's fields (e.g. one might have two clocks and DMA handler, another might only have a single clock).  That kind of use would likely require an additional API to access the stored information, however.

Basically, from I'm not really sure what's to be gained if you don't differentiate the type vs storing the clock/DMA/vendor specific in the device config.

Reply:
Why would you do a clock tree walk at runtime? (if there is a device tree to "walk" through: it's via DTS itself, never at runtime)

About the type: one always know what device "type" it is dealing with. (you don't call spi API with a watchdog device for instance,). At runtime there is no need to know the type of a the device, since all is built-in.

Reply:
Honestly?  I'd have to leave that to someone else to really defend, my examples are all contrived, that's beyond my current use cases.  The best case I could come up with would be controlling clocks/PM.  e.g. not all clocks are available in all sleep modes, so you either need to make the driver aware of what levels its clocks come from (extending the clock control API somehow) or allow the PM to walk the devices to figure out what ones are blocking the deeper sleeps based on what clocks they source; for SAM0 this is configurable, so the driver can't know what sleep levels it can operate in without introspecting the clock beyond what the current clock API would allow.

Regarding the device type, I'd expect it to be less about using the device specific API and more about using "global" features (e.g. PM/clock control) that don't necessarily depend on a specific device type (i.e. usage external to the driver implementation).

Either way, I'm not seeing a gain to an external API if the lookup is opaque: it seems like you'd just be better off storing it in the device config and never exposing it externally.